File: 687_Longest_Univalue_Path.java
/*
https://leetcode.com/problems/longest-univalue-path/
Given a binary tree, find the length of the longest path where each node in the path has the same value.
This path may or may not pass through the root.
The length of path between two nodes is represented by the number of edges between them.

Helper function return an array with the max paths via left and right children of a node. If a node has a child and
the child has the same val then it has max path via that child of 1 + max(left and right child paths).
Time - O(n)
Space - O(n)
*/

class Solution {

    int longest = 0;

    public int longestUnivaluePath(TreeNode root) {
        helper(root);
        return longest;
    }

    private int[] helper(TreeNode node) {

        int[] paths = {0, 0};
        if (node == null)
            return paths;

        int[] leftPath = helper(node.left);
        int[] rightPath = helper(node.right);
        if (node.left != null && node.left.val == node.val)
            paths[0] = 1 + Math.max(leftPath[0], leftPath[1]);
        if (node.right != null && node.right.val == node.val)
            paths[1] = 1 + Math.max(rightPath[0], rightPath[1]);

        longest = Math.max(longest, paths[0] + paths[1]);
        return paths;
    }
}

File: python_1_to_1000/687_Longest_Univalue_Path.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/longest-univalue-path/
# Given a binary tree, find the length of the longest path where each node in the path has the same value.
# This path may or may not pass through the root.
# The length of path between two nodes is represented by the number of edges between them.

# Helper function updates longest for the univalue path that goes through each node. Helper returns a tuple of the
# longest paths via the left and right children of each node.
# Time - O(n)
# Space - O(n)

class Solution(object):
    def longestUnivaluePath(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.longest = 0

        def helper(node):
            if not node:
                return 0, 0

            max_left = max(helper(node.left))       # longest univalue path from left child
            max_right = max(helper(node.right))

            # if left child has the same val as node, add one edge to longest path via left child
            left = 1 + max_left if node.left and node.left.val == node.val else 0
            right = 1 + max_right if node.right and node.right.val == node.val else 0

            self.longest = max(self.longest, left + right)

            return left, right

        helper(root)
        return self.longest